lib/maths/COneOfNPrior.cc

```diff
@@ -322,83 +363,92 @@ void COneOfNPrior::addSamples(const TDouble1Vec& samples,
+
+    // If none of the models are a good fit for the new data then restrict
+    // the maximum Bayes Factor since the data likelihood given the model
+    // is most useful if one of the models is (mostly) correct.
+    double m{std::max(n, 1.0)};
+    double maxLogBayesFactor{-m * MAXIMUM_LOG_BAYES_FACTOR};
 
-    // Compute the unnormalized posterior weights and update the component
-    // priors. These weights are computed on the side since they are only
-    // updated if all marginal likelihoods can be computed.
-    TDouble5Vec logLikelihoods;
-    TMaxAccumulator maxLogLikelihood;
-    TBool5Vec used, uses;
     for (auto& model : m_Models) {
-        bool use = model.second->participatesInModelSelection();
-
-        // Update the weights with the marginal likelihoods.
-        double logLikelihood = 0.0;
-        maths_t::EFloatingPointErrorStatus status =
-            use ? model.second->jointLogMarginalLikelihood(samples, weights, logLikelihood)
-                : maths_t::E_FpOverflowed;
-
-        if (status & maths_t::E_FpFailed) {
-            LOG_ERROR(<< "Failed to compute log-likelihood");
-            LOG_ERROR(<< "samples = " << core::CContainerPrinter::print(samples));
-            return;
-        }
 
-        if (!(status & maths_t::E_FpOverflowed)) {
-            logLikelihood += model.second->unmarginalizedParameters() * penalty;
-            logLikelihoods.push_back(logLikelihood);
-            maxLogLikelihood.add(logLikelihood);
-        } else {
-            logLikelihoods.push_back(MINUS_INF);
+        double minusBic{0.0};
+        maths_t::EFloatingPointErrorStatus status{maths_t::E_FpOverflowed};
+
+        used.push_back(model.second->participatesInModelSelection());
+        if (used.back()) {
+            double logLikelihood;
+            status = model.second->jointLogMarginalLikelihood(samples, weights, logLikelihood);
+
+            minusBic = 2.0 * logLikelihood -
+                       model.second->unmarginalizedParameters() * CTools::fastLog(n);
+
+            double modeLogLikelihood;
+            TDouble1Vec modes;
+            modes.reserve(weights.size());
+            for (const auto& weight : weights) {
+                modes.push_back(model.second->marginalLikelihoodMode(weight));
+            }
+            model.second->jointLogMarginalLikelihood(modes, weights, modeLogLikelihood);
+            maxLogBayesFactor = std::max(
+                maxLogBayesFactor, logLikelihood - std::max(modeLogLikelihood, logLikelihood));
         }
 
+        minusBics.push_back((status & maths_t::E_FpOverflowed) ? MINUS_INF : minusBic);
+        failed |= (status & maths_t::E_FpFailed);
+
         // Update the component prior distribution.
         model.second->addSamples(samples, weights);
 
-        used.push_back(use);
+        varianceMismatchPenalties.push_back(
+            -m * MAXIMUM_LOG_BAYES_FACTOR *
+            std::max(1.0 - 9.0 * CBasicStatistics::variance(m_SampleMoments) /
```

Comment:
I'm curious about the use of the `9.0` factor here - perhaps replacing with a named constant would aid understanding?

Reply:
This is the maximum relative error in the estimated variance of the model (vs the sample variance) for which we will not penalise it at all. I named this constant in [7c76c7f](https://github.com/elastic/ml-cpp/pull/468/commits/7c76c7f376c367944cea4778e61d7c28f1a9ff59).